src/iface/ethernet.rs

```diff
@@ -853,6 +875,44 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
         packet
     }
 
+    #[cfg(feature = "proto-ipv6")]
+    fn process_hopbyhop<'frame>(&mut self, sockets: &mut SocketSet, timestamp: Instant,
+                                ipv6_repr: Ipv6Repr, handled_by_raw_socket: bool,
+                                ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
+    {
+        let hbh_pkt = Ipv6HopByHopHeader::new_checked(ip_payload)?;
+        let hbh_repr = Ipv6HopByHopRepr::parse(&hbh_pkt)?;
+        let buffer_len = hbh_repr.buffer_len();
+
+        let mut i = 2;
+        while i < buffer_len {
+            let opt_pkt = Ipv6Option::new_checked(&ip_payload[i..])?;
+            let opt_repr = Ipv6OptionRepr::parse(&opt_pkt)?;
+            match opt_repr {
+                Ipv6OptionRepr::Pad1 | Ipv6OptionRepr::PadN(_) => (),
+                _ => match Ipv6OptionFailureType::from(opt_pkt.option_type()) {
+                    Ipv6OptionFailureType::Skip => (),
+                    Ipv6OptionFailureType::Discard => {
+                        return Ok(Packet::None);
+                    },
+                    _ => {
+                        // FIXME(dlrobertson): Send an ICMPv6 parameter problem message
+                        // here.
+                        return Err(Error::Unrecognized);
```

Comment:
Generating a correct Parameter Problem message we also need to keep track of the current offset.